apispec/core.py

```diff
@@ -156,15 +156,26 @@ def to_dict(self):
             ret['swagger'] = self.openapi_version.vstring
             ret['definitions'] = self._definitions
             ret['parameters'] = self._parameters
+            ret.update(self.options)
 
         elif self.openapi_version.version[0] == 3:
             ret['openapi'] = self.openapi_version.vstring
-            ret['components'] = {
-                'schemas': self._definitions,
-                'parameters': self._parameters,
-            }
+            options = self.options.copy()
+            components = options.pop('components', {})
+
+            # deep update components object
+            definitions = components.pop('schemas', {})
+            definitions.update(self._definitions)
+            parameters = components.pop('parameters', {})
+            parameters.update(self._parameters)
+
+            ret['components'] = dict(
+                schemas=definitions,
+                parameters=parameters,
+                **components
+            )
+            ret.update(options)
 
-        ret.update(self.options)
```

Comment:
`ret.update(self.options)` can be kept factorized.

Reply:
If I explicitly pass components in options, this line of code would shadow ret['components'] that is generated by codes above. And that is why I have to differentiate the behaviors between OpenAPI 2 and 3.